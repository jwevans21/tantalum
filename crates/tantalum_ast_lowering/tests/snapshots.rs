macro_rules! snapshot_from_file {
    ($file_name:ident) => {
        #[test]
        fn $file_name() {
            let source = include_str!(concat!("../../examples/", stringify!($file_name), ".ta"));
            let lexer = tantalum_lexer::Lexer::new(concat!(stringify!($file_name), ".ta"), source);

            let mut parser = tantalum_parser::Parser::new(lexer);

            let ast = parser.parse().expect("failed to parse source");

            let mut context = tantalum_ast_lowering::ASTLoweringContext::new();

            context.lower(&ast);

            let package = context.finish();

            insta::assert_debug_snapshot!(package);
        }
    };
    ($($file_name:ident),*) => {
        $(snapshot_from_file! { $file_name })*
    };
}

snapshot_from_file! {
    start,
    hello_world,
    conditionals,
    loops
}
