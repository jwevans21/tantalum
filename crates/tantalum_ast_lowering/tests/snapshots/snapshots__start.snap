---
source: tantalum_ast_lowering/tests/snapshots.rs
expression: package
snapshot_kind: text
---
HLIRPackage {
    types: {
        HLIRPath {
            components: [
                "bool",
            ],
        }: Builtin(
            Bool,
        ),
        HLIRPath {
            components: [
                "char",
            ],
        }: Builtin(
            Char,
        ),
        HLIRPath {
            components: [
                "f32",
            ],
        }: Builtin(
            F32,
        ),
        HLIRPath {
            components: [
                "f64",
            ],
        }: Builtin(
            F64,
        ),
        HLIRPath {
            components: [
                "i16",
            ],
        }: Builtin(
            I16,
        ),
        HLIRPath {
            components: [
                "i32",
            ],
        }: Builtin(
            I32,
        ),
        HLIRPath {
            components: [
                "i64",
            ],
        }: Builtin(
            I64,
        ),
        HLIRPath {
            components: [
                "i8",
            ],
        }: Builtin(
            I8,
        ),
        HLIRPath {
            components: [
                "str",
            ],
        }: Builtin(
            Str,
        ),
        HLIRPath {
            components: [
                "u16",
            ],
        }: Builtin(
            U16,
        ),
        HLIRPath {
            components: [
                "u32",
            ],
        }: Builtin(
            U32,
        ),
        HLIRPath {
            components: [
                "u64",
            ],
        }: Builtin(
            U64,
        ),
        HLIRPath {
            components: [
                "u8",
            ],
        }: Builtin(
            U8,
        ),
        HLIRPath {
            components: [
                "void",
            ],
        }: Builtin(
            Void,
        ),
    },
    scope: HLIRScope {
        next_index: ScopedValueIndex(
            1,
        ),
        names: {
            ScopedValueIndex(
                0,
            ): (
                Function(
                    HLIRFunctionPrototype {
                        name: "main",
                        inner: HLIRFunctionPrototypeAnonymous {
                            parameters: [],
                            is_variadic: false,
                            return_type: Builtin(
                                I32,
                            ),
                        },
                        parameter_names: [],
                    },
                ),
                Some(
                    Function(
                        HLIRFunctionType {
                            prototype: HLIRFunctionPrototypeAnonymous {
                                parameters: [],
                                is_variadic: false,
                                return_type: Builtin(
                                    I32,
                                ),
                            },
                        },
                    ),
                ),
            ),
        },
        global: RefCell {
            value: HLIRScopeBlock {
                variables: {
                    "main": ScopedValueIndex(
                        0,
                    ),
                },
                parent: None,
            },
        },
        frames: [],
    },
    prototypes: {
        "main": HLIRFunctionPrototype {
            name: "main",
            inner: HLIRFunctionPrototypeAnonymous {
                parameters: [],
                is_variadic: false,
                return_type: Builtin(
                    I32,
                ),
            },
            parameter_names: [],
        },
    },
    functions: {
        HLIRFunctionPrototype {
            name: "main",
            inner: HLIRFunctionPrototypeAnonymous {
                parameters: [],
                is_variadic: false,
                return_type: Builtin(
                    I32,
                ),
            },
            parameter_names: [],
        }: HLIRFunction {
            prototype: HLIRFunctionPrototype {
                name: "main",
                inner: HLIRFunctionPrototypeAnonymous {
                    parameters: [],
                    is_variadic: false,
                    return_type: Builtin(
                        I32,
                    ),
                },
                parameter_names: [],
            },
            scope: RefCell {
                value: HLIRScopeBlock {
                    variables: {},
                    parent: Some(
                        RefCell {
                            value: HLIRScopeBlock {
                                variables: {
                                    "main": ScopedValueIndex(
                                        0,
                                    ),
                                },
                                parent: None,
                            },
                        },
                    ),
                },
            },
            body: Block(
                HLIRBlock {
                    scope: RefCell {
                        value: HLIRScopeBlock {
                            variables: {},
                            parent: Some(
                                RefCell {
                                    value: HLIRScopeBlock {
                                        variables: {},
                                        parent: Some(
                                            RefCell {
                                                value: HLIRScopeBlock {
                                                    variables: {
                                                        "main": ScopedValueIndex(
                                                            0,
                                                        ),
                                                    },
                                                    parent: None,
                                                },
                                            },
                                        ),
                                    },
                                },
                            ),
                        },
                    },
                    statements: [
                        Return(
                            HLIRReturn {
                                value: Some(
                                    Literal(
                                        HLIRLiteral(
                                            Integer {
                                                value: "0",
                                                radix: 10,
                                            },
                                        ),
                                    ),
                                ),
                            },
                        ),
                    ],
                },
            ),
        },
    },
    literals: {
        Integer {
            value: "0",
            radix: 10,
        },
    },
}
